### PR TITLE
Remove internal tracking

### PR DESCRIPTION
**Pull request by davidbeckingsale**

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
LGTM

**Comment by davidbeckingsale** [comment]
LGTM

File: src/umpire/strategy/AllocationStrategy.hpp [hunk 1]
```diff
@@ -51,7 +52,11 @@ class AllocationStrategy {
    *
    * \param ptr Pointer to free.
    */
-  virtual void deallocate(void* ptr) = 0;
+  virtual void deallocate(void* ptr, std::size_t size=0) = 0;
+
+  void* allocate_internal(std::size_t bytes);
```
Comment (davidbeckingsale) [inline src/umpire/strategy/AllocationStrategy.hpp:57]: Any thoughts on the naming of these internal/impl methods?

Comment (mcfadden8) [reply src/umpire/strategy/AllocationStrategy.hpp:57]: Can they be hiddning behind pimpl?

File: src/umpire/Allocator.cpp [hunk 2]
```diff
@@ -35,7 +38,7 @@ std::size_t Allocator::getSize(void* ptr) const
 
 std::size_t Allocator::getHighWatermark() const noexcept
 {
-  return m_allocator->getHighWatermark();
+ return m_allocator->getHighWatermark();
```
Comment (kab163) [inline src/umpire/Allocator.cpp:41]:
```suggestion
  return m_allocator->getHighWatermark();
```

Comment (kab163) [reply src/umpire/Allocator.cpp:41]: I'm being picky, but I want to add back the space here lol

File: src/umpire/Allocator.inl [hunk 3]
```diff
@@ -46,7 +54,10 @@ inline void Allocator::deallocate(void* ptr)
     UMPIRE_LOG(Info, "Deallocating a null pointer (This behavior is intentionally allowed and ignored)");
     return;
   } else {
-    m_allocator->deallocate(ptr);
+    auto record = deregisterAllocation(ptr, m_allocator);
```
Comment (mcfadden8) [inline src/umpire/Allocator.inl:57]: Does this work when `m_tracking == false`?

File: src/umpire/ResourceManager.hpp [hunk 4]
```diff
@@ -108,7 +112,7 @@ class ResourceManager {
   /*!
    * \brief Construct a new Allocator.
    */
-  template <typename Strategy, bool introspection = true, typename... Args>
+  template <typename Strategy, bool introspection = false, typename... Args>
```
Comment (mcfadden8) [inline src/umpire/ResourceManager.hpp:115]: I didn't know we were changing the default for enabling introspection.  Was this intentional?

File: src/umpire/op/HostReallocateOperation.cpp [hunk 5]
```diff
@@ -45,7 +45,7 @@ void HostReallocateOperation::transform(
     }
 
     ResourceManager::getInstance().registerAllocation(
-        *new_ptr, {*new_ptr, new_size, allocator});
+        *new_ptr, {*new_ptr, new_size, new_allocation->strategy});
```
Comment (mcfadden8) [inline src/umpire/op/HostReallocateOperation.cpp:48]: Shouldn't this be conditional upon whether introspection is enabled?

File: src/umpire/resource/CudaConstantMemoryResource.cu [hunk 6]
```diff
@@ -52,33 +52,20 @@ void* CudaConstantMemoryResource::allocate(std::size_t bytes)
...
-  if (m_current_size > m_highwatermark)
-    m_highwatermark = m_current_size;
-
   UMPIRE_LOG(Debug, "(bytes=" << bytes << ") returning " << ret);
 
   return ret;
 }
 
-void CudaConstantMemoryResource::deallocate(void* ptr)
+void CudaConstantMemoryResource::deallocate(void* ptr, std::size_t size)
```
Comment (mcfadden8) [inline src/umpire/resource/CudaConstantMemoryResource.cu:60]: How does `size` get set when/if introspection is disabled?

File: src/umpire/resource/DefaultMemoryResource.hpp [hunk 7]
```diff
@@ -20,7 +20,9 @@ namespace resource {
  * allocate and deallocate memory.
  */
 template <typename _allocator>
-class DefaultMemoryResource : public MemoryResource {
+class DefaultMemoryResource : 
+  public MemoryResource
+{
```
Comment (kab163) [inline src/umpire/resource/DefaultMemoryResource.hpp:25]:
```suggestion
class DefaultMemoryResource : public MemoryResource {
```

Comment (kab163) [reply src/umpire/resource/DefaultMemoryResource.hpp:25]: (Again, being picky, but I think most of our other inheritance lines like this have been structured this way instead)

File: src/umpire/strategy/mixins/Inspector.cpp [hunk 8]
```diff
@@ -12,41 +12,34 @@ namespace umpire {
...
 void
 Inspector::registerAllocation(
     void* ptr,
     std::size_t size,
-    strategy::AllocationStrategy* strategy)
+    strategy::AllocationStrategy* s) 
 {
-  m_current_size += size;
-  m_allocation_count++;
+  s->m_current_size += size;
```
Comment (mcfadden8) [inline src/umpire/strategy/mixins/Inspector.cpp:21]: Directly changing state of another object seems sneaky to me.  I appreciate why it is being done this way, but is there perhaps a different way to accomplish the same thing without having to worry about the strategy object and inspector object from inadvertantly overwriting each other?

File: src/umpire/util/Macros.hpp [hunk 9]
```diff
@@ -143,4 +143,12 @@
...
 #endif
 
+#if 0
+#define UMPIRE_INTERNAL_TRACK(p, s) registerAllocation(p, s, this);
+#define UMPIRE_INTERNAL_UNTRACK(p) deregisterAllocation(p, this);
+#else 
+#define UMPIRE_INTERNAL_TRACK(p, s)
+#define UMPIRE_INTERNAL_UNTRACK(p) umpire::util::AllocationRecord{};
+#endif
+
```
Comment (mcfadden8) [inline src/umpire/util/Macros.hpp:153]: Are these still needed?

File: tests/unit/resource/resource_tests.cpp [hunk 10]
```diff
@@ -46,5 +46,5 @@ TEST(DefaultMemoryResource, AllocateDeallocate)
   double* pointer = (double*)alloc->allocate(10 * sizeof(double));
   ASSERT_NE(pointer, nullptr);
 
-  alloc->deallocate(pointer);
+  alloc->deallocate(pointer, 0);
```
Comment (mcfadden8) [inline tests/unit/resource/resource_tests.cpp:49]: Why isn't this `alloc->deallocate(pointer);`?

**Review by mcfadden8 — CHANGES_REQUESTED** [review]
Overall, this looks good and I really like how things seem to have simplified.  I'm still not sure I follow how things will work for Allocators that do not have tracking enabled which is the source of a lot of my questions.

File: tests/unit/resource/resource_tests.cpp (hunk 10 above)
Comment (mcfadden8) [inline tests/unit/resource/resource_tests.cpp:49]: What are the semantics of `size=0`?  Are they documented somewhere?

File: src/umpire/resource/MemoryResource.hpp [hunk 11]
```diff
@@ -53,28 +53,7 @@ class MemoryResource : public strategy::AllocationStrategy {
...
-   * \brief Return the memory high watermark for this MemoryResource.
-   *
-   * This is the largest amount of memory allocated by this Allocator. Note
-   * that this may be larger than the largest value returned by
-   * getCurrentSize.
-   *
-   * \return Memory high watermark.
-   */
-  virtual std::size_t getHighWatermark() const noexcept override = 0;
+  virtual void deallocate(void* ptr, std::size_t size) override = 0;
```
Comment (mcfadden8) [inline src/umpire/resource/MemoryResource.hpp:56]: Please add documentation for `size` parameter above.  Specifically, the semantics for when `size == 0`

**Review by mcfadden8 — COMMENTED** [review]
Looks great.  I've reviewed and added comments and some suggestions (mostly minor).

File: benchmarks/fixed_pool_stress_test.cpp [hunk 12]
```diff
@@ -12,6 +12,7 @@
 #include "umpire/config.hpp"
 #include "umpire/ResourceManager.hpp"
 #include "umpire/Allocator.hpp"
+#include "umpire/strategy/FixedPool.hpp"
```
Comment (kab163) [inline benchmarks/fixed_pool_stress_test.cpp:15]: Why was this added?

**Review by mcfadden8 — APPROVED** [review]
Looks good.  Thanks.
